src/node_jsvmapi.cc

```diff
@@ -1569,122 +1600,153 @@ napi_status napi_coerce_to_string(napi_env e, napi_value v, napi_value* result)
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_wrap(napi_env e, napi_value jsObject, void* nativeObj,
-                      napi_destruct destructor, napi_weakref* handle) {
+napi_status napi_wrap(napi_env e,
+                      napi_value jsObject,
+                      void* nativeObj,
+                      napi_finalize finalize_cb,
+                      napi_ref* result) {
   NAPI_PREAMBLE(e);
+  CHECK_ARG(jsObject);
+  CHECK_ARG(result);
 
   v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
-  v8::Local<v8::Context> context = isolate->GetCurrentContext();
-  v8::Local<v8::Object> obj;
-  CHECK_TO_OBJECT(context, obj, jsObject);
 
-  v8impl::ObjectWrapWrapper* wrap =
-      new v8impl::ObjectWrapWrapper(obj, nativeObj, destructor);
+  v8::Local<v8::Object> obj = v8impl::V8LocalValueFromJsValue(jsObject).As<v8::Object>();
+  assert(obj->InternalFieldCount() > 0);
+  obj->SetAlignedPointerInInternalField(0, nativeObj);
 
-  if (handle)
-  {
-    return napi_create_weakref(
-      e,
-      v8impl::JsValueFromV8LocalValue(wrap->handle()),
-      handle);
-  }
+  v8impl::Reference* reference = new v8impl::Reference(
+    isolate, obj, 0, false, finalize_cb, nativeObj);
+  *result = reinterpret_cast<napi_ref>(reference);
 
-  // TODO: Is the handle parameter really optional?
-  //       Why would anyone want to construct an object wrap and immediately lose it?
   return GET_RETURN_STATUS();
 }
 
 napi_status napi_unwrap(napi_env e, napi_value jsObject, void** result) {
   NAPI_PREAMBLE(e);
+  CHECK_ARG(jsObject);
   CHECK_ARG(result);
 
-  v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
-  v8::Local<v8::Context> context = isolate->GetCurrentContext();
-  v8::Local<v8::Object> obj;
-  CHECK_TO_OBJECT(context, obj, jsObject);
+  v8::Local<v8::Object> obj = v8impl::V8LocalValueFromJsValue(jsObject).As<v8::Object>();
+
+  assert(obj->InternalFieldCount() > 0);
```

Comment:
I'd presume the following call to `GetAlignedPointerFromInternalField(0)` should fail if the internal field count is 0. But fail how? Would it set a JavaScript exception? If so, `GET_RETURN_STATUS()` will detect that. I still don't have a good understanding of V8's error behavior, aside from maybes. Like, what if the value passed in is a primitive, or is undefined? (I remove the coercion to object because if it wasn't an object to start with then coercing to an object is not going to help with unwrapping anything.)